documentation/ArrayComposition.md

```diff
@@ -0,0 +1,130 @@
+FORALL(J=1:N,K=1:M) T(J,K)=A(J,K) + 1.0
+FORALL(J=1:N,K=1:M) B(J,K)=T(J,K)
+DEALLOCATE(T(N,M))
+```
+But we can avoid the allocation, population, and deallocation of
+the temporary by treating the right-hand side expression as if it
+were a statement function `F(J,K)=A(J,K)+1.0` and evaluating
+```
+FORALL(J=1:N,K=1:M) A(J,K)=F(J,K)
+```
+
+In general, when a Fortran array assignment to a non-allocatable array
+does not include the left-hand
+side variable as an operand of the right-hand side expression, and any
+function calls on the right-hand side are elemental or scalar-valued,
+we can avoid the use of a temporary.
+
+Transformational intrinsic functions as function composition
+============================================================
+Many of the transformational intrinsic functions listed above
+can, when their array arguments are viewed as functions over their
+index tuples, be seen as compositions of those functions with
+functions of the "incoming" indices.
+
+For example, the application of `TRANSPOSE(A + 1.0)` to the index
+tuple `(J,K)` becomes `A(K,J) + 1.0`.
+
+Determination of rank and shape
+===============================
+An important part of evaluating array expressions without the use of
+temporary storage is determining the shape of the result prior to,
+or without, evaluating the elements of the result.
+
+The shapes of array objects, results of elemental intrinsic functions,
+and results of intrinsic operations are obvious.
+But it is possible to determine the shapes of the results of many
+transformantional intrinsic function calls as well.
+
+* `SHAPE(SUM(X,DIM=d))` is `SHAPE(X)` with one element removed.
+  The `DIM=` argument is commonly a compile-time constant.
+* `SHAPE(MAXLOC(X))` is `[RANK(X)]`.
+* `SHAPE(MAXLOC(X,DIM=d))` is `SHAPE(X)` with one element removed.
+* `SHAPE(TRANSPOSE(M))` is a reversal of `SHAPE(M)`.
+* `SHAPE(RESHAPE(..., SHAPE=S))` is `S`.
+* `SHAPE(CSHIFT(X))` is `SHAPE(X)`; same with `EOSHIFT`.
+* `SHAPE(PACK(A,VECTOR=V))` is `SHAPE(V)`; `RANK(PACK(...))` is always 1.
+* `SHAPE(UNPACK(MASK=M))` is `SHAPE(M)`.
+* `SHAPE(SHAPE(X))` is `[RANK(X)]`.
+
+This is useful because expression evaluations that *do* require temporaries
+to hold their results (due to the context in which the evaluation occurs)
+can be implemented with a separation of the allocation
+of the temporary array and the population of the array.
+The code that evaluates the expression, or that implements a transformational
+intrinsic in the runtime library, can be designed with an API that includes
+a pointer to the destination array as an argument.
+
+Statements like `ALLOCATE(A,SOURCE=expression)` should thus be capable
+of evaluating their array expressions directly into the newly-allocated
+storage for the allocatable array.
+The implementation would generate code to calculate the shape, use it
+to allocate the memory and populate the descriptor, and then drive a
+loop nest around the expression to populate the array.
```

Comment:
Yes, I do think that the compiler should create the loop nest; it's basically a `FORALL`.  This would enable program restructuring transformations, loop optimizations, multi-version code generation, automatic GPU acceleration, &c.